From a0943bd277ed9b5ecaf1c7c1ace7590b1b6da37d Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH] ExifInterface: fix incorrect thumbnailLength in handleThumbnailFromJfif

Incorrect thumbnailLength in ExifInterface::handleThumbnailFromJfif
potentially leads to incomplete bytes of thumbnail.

The statement below potentially leads to incorrect thumbnailLength
caused by in.available.

thumbnailLength = Math.min(thumbnailLength, in.available() - thumbnailOffset);

Test: Invoke ExifInterface.getThumbnail to get thumbnail data stored in jpeg file.

diff --git a/media/java/android/media/ExifInterface.java b/media/java/android/media/ExifInterface.java
index bf3a3b9b07b..fed91c2b1dc 100644
--- a/media/java/android/media/ExifInterface.java
+++ b/media/java/android/media/ExifInterface.java
@@ -3203,7 +3203,7 @@ private void handleThumbnailFromJfif(ByteOrderedDataInputStream in, HashMap thum
             int thumbnailLength = jpegInterchangeFormatLengthAttribute.getIntValue(mExifByteOrder);
 
             // The following code limits the size of thumbnail size not to overflow EXIF data area.
-            thumbnailLength = Math.min(thumbnailLength, in.available() - thumbnailOffset);
+            thumbnailLength = Math.min(thumbnailLength, in.mLength - thumbnailOffset);
             if (mMimeType == IMAGE_TYPE_JPEG || mMimeType == IMAGE_TYPE_RAF
                     || mMimeType == IMAGE_TYPE_RW2) {
                 thumbnailOffset += mExifOffset;
